File: src/test/java/pl/robert/project/bank/account/BankAccountFacadeTest.java
package pl.robert.project.bank.account;

import org.junit.Test;
import org.junit.jupiter.api.Assertions;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.validation.BindingResult;
import org.springframework.validation.DataBinder;
import pl.robert.project.bank.account.dto.ModifyBalanceDTO;
import pl.robert.project.bank.account.exception.UpdateMoneyException;
import pl.robert.project.transactions.dto.TransactionDTO;

@RunWith(SpringRunner.class)
@SpringBootTest
public class BankAccountFacadeTest {

    @Autowired
    private BankAccountFacade facade;

    @Test
    public void shouldCreateBankAccountAndSaveIntoDb() {
        // Should generate new account with random number
        BankAccount account = facade.generateBankAccount();

        // Should save new account
        facade.save(account);

        // Should find it by id and it won't be a null
        Assertions.assertNotNull(facade.findById(account.getId()));

        // Should find it by account number and it won't be a null
        Assertions.assertNotNull(facade.findByNumber(account.getNumber()));
    }

    @Test
    public void shouldUpdateMoneySuccessfully() {
        // Should create sender and receiver bank accounts
        BankAccount sender = facade.generateBankAccount();
        BankAccount receiver = facade.generateBankAccount();

        // Should set sender balance to 3000 and receiver to 1000
        sender.setBalance(3000);
        receiver.setBalance(1000);

        // Should save 2 instances
        facade.save(sender);
        facade.save(receiver);

        // Should set sender balance to 2500 and receiver to 1500
        try {
            facade.updateMoney(500, sender.getId(), receiver.getId());
        } catch (UpdateMoneyException e) {
            // This exception won't occur
        }

        // Should check if both balances has been updated
        Assertions.assertEquals(2500, facade.findById(sender.getId()).getBalance(), 0.01);
        Assertions.assertEquals(1500, facade.findById(receiver.getId()).getBalance(), 0.01);
    }

    @Test(expected = UpdateMoneyException.class)
    public void shouldNotUpdateMoneyAndThrowAnException() throws UpdateMoneyException {
        // Should create sender and receiver bank accounts
        BankAccount sender = facade.generateBankAccount();
        BankAccount receiver = facade.generateBankAccount();

        // Should save 2 instances
        facade.save(sender);
        facade.save(receiver);

        // Should throw an exception cuz sender balance is less than 0
        try {
            facade.updateMoney(500, sender.getId(), receiver.getId());
        } catch (UpdateMoneyException e) {
            throw new UpdateMoneyException();
        }
    }

    @Test
    public void shouldModifyBalanceSuccessfully() {
        // Should create new bank account and DTO
        BankAccount account = facade.generateBankAccount();
        ModifyBalanceDTO dto = new ModifyBalanceDTO();

        // Should save created bank account
        facade.save(account);

        // Should set id and new balance to 1000
        dto.setId(account.getId());
        dto.setNewBalance(1000);

        // Should modify balance to 1000
        facade.modifyBalance(dto);

        // Should check if balance has been modified
        Assertions.assertEquals(1000, facade.findById(account.getId()).getBalance());
    }

    @Test
    public void shouldCheckReceiverBankAccountNumberWithoutAndWithErrors() {
        // Should create 2 bank accounts and transaction DTO
        BankAccount sender = facade.generateBankAccount();
        BankAccount receiver = facade.generateBankAccount();
        TransactionDTO dto = new TransactionDTO();

        // Should save created bank accounts
        facade.save(sender);
        facade.save(receiver);

        // Should set fields in DTO object
        dto.setSenderAccountNumber(sender.getNumber());
        dto.setReceiverAccountNumber(receiver.getNumber());

        // Should create new BindingResult
        BindingResult result = createBindingResult(dto);

        // Should validate receiver bank account number with no errors
        facade.checkReceiverBankAccountNumber(sender.getId(), dto, result);
        Assertions.assertFalse(result.hasErrors());

        // Should validate again,
        // set sender and receiver number to sender number,
        // should reject an error
        dto.setReceiverAccountNumber(dto.getSenderAccountNumber());
        facade.checkReceiverBankAccountNumber(sender.getId(), dto, result);
        Assertions.assertTrue(result.hasErrors());
    }

    @Test
    public void shouldCheckSenderAmountWithoutAndWithErrors() {
        // Should create bank account and transaction DTO
        BankAccount sender = facade.generateBankAccount();
        TransactionDTO dto = new TransactionDTO();

        // Should set bank account balance to 1000
        sender.setBalance(1000);

        // Should save created bank account
        facade.save(sender);

        // Should set amount in DTO object
        dto.setAmount(900.0);

        // Should create new BindingResult
        BindingResult result = createBindingResult(dto);

        // Should validate sender amount with no errors
        facade.checkSenderAmount(sender.getId(), dto, result);
        Assertions.assertFalse(result.hasErrors());

        // Should validate again,
        // set dto amount to more than 1000 which is current account balance,
        // should reject an error
        dto.setAmount(1001.0);
        facade.checkSenderAmount(sender.getId(), dto, result);
        Assertions.assertTrue(result.hasErrors());
    }

    private BindingResult createBindingResult(Object obj) {
        return new DataBinder(obj).getBindingResult();
    }
}

File: src/test/kotlin/pl.robert.project/bank/account/BankAccountFacadeTest.kt
package pl.robert.project.bank.account

import org.junit.Test
import org.junit.jupiter.api.Assertions
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.junit4.SpringRunner
import org.springframework.validation.BindingResult
import org.springframework.validation.DataBinder
import pl.robert.project.bank.account.dto.ModifyBalanceDTO
import pl.robert.project.bank.account.exception.UpdateMoneyException
import pl.robert.project.transactions.dto.TransactionDTO

@RunWith(SpringRunner::class)
@SpringBootTest
class BankAccountFacadeTest {

    @Autowired
    private lateinit var facade: BankAccountFacade

    @Test fun `Should create bank account and save into db`() {
        // Generate new account with random number
        val account = facade.generateBankAccount()

        // Save new account
        facade.save(account)

        // Find it by id and it won't be a null
        Assertions.assertNotNull(facade.findById(account.id))

        // Find it by account number and it won't be a null
        Assertions.assertNotNull(facade.findByNumber(account.number))
    }

    @Test fun `Should update money successfully`() {
        // Create sender and receiver bank accounts
        val sender = facade.generateBankAccount()
        val receiver = facade.generateBankAccount()

        // Set sender balance to 3000 and receiver to 1000
        sender.balance = 3000.0
        receiver.balance = 1000.0

        // Save 2 instances
        facade.save(sender)
        facade.save(receiver)

        // Set sender balance to 2500 and receiver to 1500
        try {
            facade.updateMoney(500.0, sender.id, receiver.id)
        } catch (e: UpdateMoneyException) {
            // This exception won't occur
        }

        // Check if both balances has been updated
        Assertions.assertEquals(2500.0, facade.findById(sender.id).balance, 0.01)
        Assertions.assertEquals(1500.0, facade.findById(receiver.id).balance, 0.01)
    }

    @Test(expected = UpdateMoneyException::class)
    @Throws(UpdateMoneyException::class)
    fun `Should not update money and throw an exception`() {
        // Create sender and receiver bank accounts
        val sender = facade.generateBankAccount()
        val receiver = facade.generateBankAccount()

        // Save 2 instances
        facade.save(sender)
        facade.save(receiver)

        // Throw an exception cuz sender balance is less than 0
        try {
            facade.updateMoney(500.0, sender.id, receiver.id)
        } catch (e: UpdateMoneyException) {
            throw UpdateMoneyException()
        }

    }

    @Test fun `Should modify balance successfully`() {
        // Create new bank account and DTO
        val account = facade.generateBankAccount()
        val dto = ModifyBalanceDTO()

        // Save created bank account
        facade.save(account)

        // Set id and new balance to 1000
        dto.id = account.id
        dto.newBalance = 1000.0

        // Modify balance to 1000
        facade.modifyBalance(dto)

        // Check if balance has been modified
        Assertions.assertEquals(1000.0, facade.findById(account.id).balance)
    }

    @Test fun `Should check receiver bank account number without and with errors`() {
        // Create 2 bank accounts and transaction DTO
        val sender = facade.generateBankAccount()
        val receiver = facade.generateBankAccount()
        val dto = TransactionDTO()

        // Save created bank accounts
        facade.save(sender)
        facade.save(receiver)

        // Set fields in DTO object
        dto.senderAccountNumber = sender.number
        dto.receiverAccountNumber = receiver.number

        // Create new BindingResult
        val result = createBindingResult(dto)

        // Validate receiver bank account number with no errors
        facade.checkReceiverBankAccountNumber(sender.id, dto, result)
        Assertions.assertFalse(result.hasErrors())

        // Set sender and receiver number to sender number,
        // Validate receiver bank account number,
        // Reject an error
        dto.receiverAccountNumber = dto.senderAccountNumber
        facade.checkReceiverBankAccountNumber(sender.id, dto, result)
        Assertions.assertTrue(result.hasErrors())
    }

    @Test fun `Should check sender amount without and with errors`() {
        // Create bank account and transaction DTO
        val sender = facade.generateBankAccount()
        val dto = TransactionDTO()

        // Set bank account balance to 1000
        sender.balance = 1000.0

        // Save created bank account
        facade.save(sender)

        // Set amount in DTO object
        dto.amount = 900.0

        // Create new BindingResult
        val result = createBindingResult(dto)

        // Validate sender amount with no errors
        facade.checkSenderAmount(sender.id, dto, result)
        Assertions.assertFalse(result.hasErrors())

        // Set dto amount to 1001 which is more than current balance,
        // Validate sender amount,
        // Reject an error
        dto.amount = 1001.0
        facade.checkSenderAmount(sender.id, dto, result)
        Assertions.assertTrue(result.hasErrors())
    }

    private fun createBindingResult(obj: Any): BindingResult {
        return DataBinder(obj).bindingResult
    }
}
